File: simple-oop-java/src/main/java/hvm/Hero.java
package hvm;

/**
 * Hero is a GameCharacter that can fight Monsters.
 */
public class Hero extends GameCharacter {

	public Hero(String name, int health, int attack, int speed) {
		super(name, health, attack, speed);
	}

	/**
	 * Fights a monster.
	 * How do we know if behaviour belongs in Hero or Monster or GameCharacter?
	 */
	public void fight(Monster monster) {
		if (monster.getSpeed() > getSpeed()) {
			monster.hit(this);
			if (isAlive()) {
				hit(monster);
			}
		} else {
			hit(monster);
			if (monster.isAlive()) {
				monster.hit(this);
			}
		}
	}

}

File: c-sharp-extensions/src/Hero.cs
namespace HeroVsMonsters;

public record struct Hero(string Name, int Health, int Attack, int Speed)
{
    public bool IsAlive => Health > 0;

    public bool IsDead => Health <= 0;

    private void Hit(Monster target) => target.TakeHit(Attack);

    public void TakeHit(int attack, int speedDamage)
    {
        Health -= attack;
        Speed -= speedDamage;
    }

    public void UseItem(Item item)
    {
        Health += item.Health;
        Attack += item.Attack;
        Speed += item.Speed;
    }

    public void Fight(Monster monster)
    {
        if (monster.Speed > Speed)
        {
            monster.Hit(this);
            if (IsAlive)
            {
                Hit(monster);
            }
        }
        else
        {
            Hit(monster);
            if (monster.IsAlive)
            {
                monster.Hit(this);
            }
        }
    }
}